zerver/openapi/zulip.yaml

```diff
@@ -6622,25 +6633,39 @@ paths:
           content:
             application/json:
               schema:
-                allOf:
-                  - $ref: "#/components/schemas/CodedError"
-                  - properties:
-                      msg:
-                        enum:
-                          - Your organization has turned off message editing
-                          - You don't have permission to edit this message
-                          - The time limit for editing this message has past
-                          - Nothing to change
-                          - Topic can't be empty
-                    example:
-                      {
-                        "code": "BAD_REQUEST",
-                        "msg": "You don't have permission to edit this message",
-                        "result": "error",
-                      }
-                    description: |
-                      A typical JSON response for when one doesn't have the permission to
-                      edit a particular message:
+                oneOf:
+                  - allOf:
+                      - $ref: "#/components/schemas/CodedError"
+                      - properties:
+                          msg:
+                            enum:
+                              - Your organization has turned off message editing
+                              - You don't have permission to edit this message
+                              - The time limit for editing this message has past
+                              - Nothing to change
+                              - Topic can't be empty
+                        example:
+                          {
+                            "code": "BAD_REQUEST",
+                            "msg": "You don't have permission to edit this message",
+                            "result": "error",
+                          }
+                        description: |
+                          A typical JSON response for when one doesn't have the permission to
+                          edit a particular message:
+                  - allOf:
+                      - $ref: "#/components/schemas/CodedError"
+                      - example:
+                          {
+                            "code": "MESSAGE_MOVE_TIME_LIMIT_EXCEEDED",
```

Comment:
Agreed that we should do this, though I'm OK with it being a follow-up just to get all the other complexity of this change into `main`.

Reply:
I ended up working on this despite my previous comment, here's the diff:

``` diff
                       - example:
                           {
                             "code": "MESSAGE_MOVE_TIME_LIMIT_EXCEEDED",
+                            "first_message_id_allowed_to_move": 123,
                             "msg": "Can't move all messages in this topic. You can only move 2/5 latest messages",
                             "result": "error",
+                            "total_messages_allowed_to_move": 2,
+                            "total_messages_in_topic": 5,
                           }
                         description: |
-                          A typical failed JSON response for when time limit for moving messages has
-                          passed.
+                          A special failed JSON response for when the user has permission to move
+                          the target message, but asked to move all messages in a topic
+                          (`change_all`) and the user does not have permission to move the entire
+                          topic.
+
+                          This happens when the topic contains some messages that are older than an
+                          applicable time limit for the requested topic move
+                          (`move_messages_within_stream_limit_seconds` and/or
+                          `move_messages_between_streams_limit_seconds`).
+
+                          This response contains data on which portion of this topic the user has
+                          permission to move; `first_message_id_allowed_to_move` is the oldest
+                          message ID in this topic that the user has permission to move; the user
+                          can move `total_messages_allowed_to_move` messages, and
+                          `total_messages_in_topic` are outside the user's permissions.
+
+                          A client is recommended to either just present the error to the user, or
+                          prompt the user with this information. In such a prompt, clients should
+                          offer the option of cancelling and (manually) asking a moderator to move
+                          the entire topic, since that's often a better experience than partially
+                          moving a topic. This API supports a client that wishes to allow the user
+                          to repeat the request with a `change_later` propagation mode and a target
+                          message ID of `first_message_id_allowed_to_move`, if desired.
+
+                          Note that in a stream with protected history, the Zulip security model
+                          requires that the above calculations only include messages the acting user
+                          has access to.
 
                           **Changes**: New in Zulip 7.0 (feature level 172).
     delete:
```